js/src/components/paid-ads/convertToAssetGroupUpdateBody.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * Internal dependencies
+ */
+import { ASSET_KEY, ASSET_FORM_KEY } from '.~/constants';
+
+/**
+ * @typedef {import('.~/components/types.js').AssetGroupFormValues} AssetGroupFormValues
+ * @typedef {import('.~/data/types.js').AssetEntityGroup} AssetEntityGroup
+ * @typedef {import('.~/data/types.js').AssetOperations} AssetOperations
+ * @typedef {import('.~/data/types.js').AssetEntityGroupUpdateBody} AssetEntityGroupUpdateBody
+ */
+
+/**
+ * Differences the assets form values against the existing asset entity group and
+ * returns the updating operations of the asset group.
+ *
+ * @param  {AssetEntityGroup} assetGroup The existing asset entity group as the base data for comparing the differences.
+ * @param  {AssetGroupFormValues} values The assets form values to be compared.
+ * @return {AssetOperations[]} The creation and deletion operations for updating the asset group.
+ */
+export function diffAssetOperations( assetGroup, values ) {
+	const operations = [];
+	const assets = assetGroup.assets;
+
+	function pushDeletions( key, ...assetEntities ) {
+		const deletions = assetEntities.map( ( assetEntity ) => ( {
+			...assetEntity,
+			// Setting the `content` to null indicates the asset deletion operation.
+			content: null,
+			field_type: key,
+		} ) );
+		operations.push( ...deletions );
+	}
+
+	Object.values( ASSET_KEY ).forEach( ( key ) => {
+		const contents = [ values[ key ] ].flat().filter( Boolean );
+		const assetEntities = [ assets[ key ] ].flat().filter( Boolean );
+		let entityIndex = 0;
+
+		contents.forEach( ( content ) => {
+			do {
+				const assetEntity = assetEntities[ entityIndex ];
+
+				if ( content === assetEntity?.content ) {
+					break;
+				}
+
+				if ( assetEntity ) {
+					pushDeletions( key, assetEntity );
+				}
+
+				entityIndex += 1;
+			} while ( entityIndex < assetEntities.length );
+
+			if ( entityIndex >= assetEntities.length ) {
+				operations.push( {
+					// Setting the `id` to null indicates the asset creation operation.
+					id: null,
+					content,
+					field_type: key,
+				} );
+			}
+
+			entityIndex += 1;
+		} );
+
+		pushDeletions( key, ...assetEntities.slice( entityIndex ) );
+	} );
+
+	return operations;
+}
+
+/**
+ * Converts the assets form values into the request body based on an existing asset entity group
+ * for updating that existing asset group.
+ *
+ * @param  {AssetEntityGroup} assetGroup The existing asset entity group as the base data for comparing the differences.
+ * @param  {AssetGroupFormValues} values The assets form values to be converted.
+ * @return {AssetEntityGroupUpdateBody} The request body for updating the asset group.
+ */
+export default function convertToAssetGroupUpdateBody( assetGroup, values ) {
```

Comment:
It's declared in the docs, but still `formValues` would be more clear than just `values`.

Reply:
I would agree with it if the type of value was just declared `Object`.

However, just because JSDoc has explicitly pointed out the type and use of `values`, a relatively short name was used for this small function. Otherwise, if no matter how clear the JSDos it's, the parameter name should still very align with the explicit stuff, then the next question might be "why `formValues` but not `assetsFormValues` or `assetGroupFormValues`?"

Reply:
> "why formValues but not assetsFormValues or assetGroupFormValues?"

Yes, that would be ideal 

assetGroup --> existingAssetGroup or currentAssetGroup
values --> assetsGroupFormValues 

I don't think is small function (50 lines).

Reply:
> Yes, that would be ideal
> assetGroup --> existingAssetGroup or currentAssetGroup
> values --> assetsGroupFormValues

Sorry, but it's too wordy for me.

Things like this are probably very subjective, so I don't really have a good idea to make all devs happy with the naming. In this kind of case, I usually think if it could be understood after a few seconds (or a few minutes against those very complex things), then the naming is good enough.

> I don't think is small function (50 lines).

Not sure what I missed but this function should have only 11 lines of code.